internal/topology/upgrade/clusterctl_upgrade_test.go

```diff
@@ -0,0 +1,1176 @@
+
+	cpInfraRef, err := contract.ControlPlane().MachineTemplate().InfrastructureRef().Get(refObj)
+	g.Expect(err).To(Succeed())
+	refObj, err = getReferencedObject(ctx, env.GetClient(), cpInfraRef, "v1beta2", cluster.Namespace)
+	g.Expect(err).To(Succeed())
+	controlPlaneInfraRef := clusterv1.ContractVersionedObjectReference{
+		APIGroup: refObj.GroupVersionKind().Group,
+		Kind:     refObj.GetKind(),
+		Name:     refObj.GetName(),
+	}
+	g.Expect(refsBefore).ToNot(HaveKey(controlPlaneInfraRef), "controlPlane.spec.machineTemplate.infrastructureRef did not rollout")
+	g.Expect(refsAfter[controlPlaneInfraRef]).To(Equal(int64(1)), "controlPlane.spec.machineTemplate.infrastructureRef has unexpected generation")
+
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	err = env.List(ctx, machineDeployments, client.InNamespace(cluster.Namespace), client.MatchingLabels{
+		clusterv1.ClusterNameLabel:          cluster.Name,
+		clusterv1.ClusterTopologyOwnedLabel: "",
+	})
+	g.Expect(err).To(Succeed())
+	g.Expect(machineDeployments.Items).To(HaveLen(1))
+
+	for _, md := range machineDeployments.Items {
+		mdRef := clusterv1.ContractVersionedObjectReference{
+			APIGroup: clusterv1.GroupVersion.Group,
+			Kind:     "MachineDeployment",
+			Name:     md.GetName(),
+		}
+		g.Expect(refsAfter[mdRef]).To(Equal(refsBefore[mdRef]+1), "machineDeployment "+md.Name+" unexpected change") // MachineDeployment is expected to have an additional generation due to rollout
+
+		refObj, err = getReferencedObject(ctx, env.GetClient(), &md.Spec.Template.Spec.InfrastructureRef, "v1beta2", cluster.Namespace)
+		g.Expect(err).To(Succeed())
+		mdInfraRef := clusterv1.ContractVersionedObjectReference{
+			APIGroup: refObj.GroupVersionKind().Group,
+			Kind:     refObj.GetKind(),
+			Name:     refObj.GetName(),
+		}
+		g.Expect(refsBefore).ToNot(HaveKey(mdInfraRef), "machineDeployment "+md.Name+" .spec.template.spec.infrastructureRef did not rollout")
+		g.Expect(refsAfter[mdInfraRef]).To(Equal(int64(1)), "machineDeployment "+md.Name+" .spec.template.spec.infrastructureRef has unexpected generation")
+
+		refObj, err = getReferencedObject(ctx, env.GetClient(), md.Spec.Template.Spec.Bootstrap.ConfigRef, "v1beta2", cluster.Namespace)
+		g.Expect(err).To(Succeed())
+		mdBootstrapRef := clusterv1.ContractVersionedObjectReference{
+			APIGroup: refObj.GroupVersionKind().Group,
+			Kind:     refObj.GetKind(),
+			Name:     refObj.GetName(),
+		}
+		g.Expect(refsBefore).ToNot(HaveKey(mdBootstrapRef), "machineDeployment "+md.Name+" .spec.template.spec.bootstrap.configRef did not rollout")
+		g.Expect(refsAfter[mdBootstrapRef]).To(Equal(int64(1)), "machineDeployment "+md.Name+" .spec.template.spec.bootstrap.configRef has unexpected generation")
+	}
+}
+
+// assertNoRollout assert a rollout did not happened by checking referenced template or referencedObjects did not changed/increased generation.
+//
+//nolint:unused
```

Comment:
```suggestion
//nolint:unused
```

If so, why do we keep this func ?

Reply:
I would prefer to keep this around because it could be useful in case I want to use this test to validate impacts on some API change